designs/autoscaler.md

```diff
@@ -0,0 +1,515 @@
+# Autoscaling on EKS Anywhere Cluster
+
+## Introduction
+
+**Problem:** EKS Anywhere (EKS-A) is currently incompatible with most of the Kubernetes cluster autoscaling products on the market. Installing [Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) in an EKS-A cluster will not work out of the box without API and design changes.
+
+EKS Anywhere builds on top of Cluster API (CAPI) and maintains several CAPI cloud providers for cluster provisioning and management.[Cluster Autoscaler on Cluster API](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider/clusterapi) was chosen as the default EKS-A cluster autoscaling solution for its native CAPI support and easy integration, which avoids the need to drastically change the API definition and workflow. The following discussion will focus on Cluster Autoscaler as the autoscaling tool to integrate into EKS Anywhere.
+
+At launch, EKS Anywhere user can specify a fixed number of nodes for each worker node group in the cluster. Once a cluster is created, EKS-A cluster controller watches this configuration for changes, generates CAPI resources, and re-applies them in the cluster. Since EKS Anywhere does not support user interaction with CAPI directly, most manual CAPI resource changes made by the user will be overwritten by the EKS-A cluster controller. This behavior prevents users from enabling Cluster Autoscaler themselves, since the required autoscaling annotations manually applied to CAPI resources will be removed by EKS-A cluster controller.
+
+This document aims to solve this limitation and provides **autoscaling** **compatibility** in EKS Anywhere, so the user can either choose to install Cluster Autoscaler as an EKS-A curated package during or after cluster creation time, or install themselves from [upstream](https://github.com/kubernetes/autoscaler/releases) after a cluster is created.
+
+### Goals and Objectives
+
+As an EKS Anywhere user, I want to:
+
+* create clusters in a streamlined fashion with autoscaling capabilities
+* add autoscaling capabilities to existing EKS-A clusters
+* see my clusters to auto scale up and down worker nodes within the boundaries I limited
+
+### Statement of Scope
+
+#### In Scope
+
+* Include Cluster Autoscaler as an [EKS-A curated package](https://anywhere.eks.amazonaws.com/docs/concepts/packages/).
+* Integrate Cluster Autoscaler to all CAPI providers since the feature is provider agnostic.
+* Integrate autoscaling configuration in cluster spec to make EKS-A cluster autoscaling-aware.
```

Comment:
Why would the autoscaling configuration be in the cluster spec?

Reply:
One major reason is that we want to guarantee user to be able to install autoscaler from upstream, without curated package. Exposing the machinedeployment annotations to cluster spec so that user does not need to interact w CAPI objects directly and eks-a cluster controller won't revert changes.

Reply:
That is true. Happy to discuss if there is a better way to support upstream cluster autoscaler without changing our cluster spec. A couple of ways to make it less confusing:
1. add warning in cli log if min/max specified but autoscaler controller is not installed
2. document the whole autoscaler installation process
3. default to install autoscaler curated pkg when detecting min/max in cluster spec during cluster creation (tho this approach introduce another problem which is -- should we auto install curated pkg w/o `--install-packages` arg, since it can be a gated feature)

Reply:
I had a similar thought. It might be useful to think through an independent CRD and controller that targets worker node groups using label selectors, identifies associated objects and applies the annotations. 

I also thought about this in the inverse. I install cluster auto scaler because we're saying customers can do that as its supported. After getting it working having applied annotations, scaling stops working due to a different unrelated change - something's seems a miss here with the experience. The above solution would at least identify missing annotations and apply them again re-enabling auto scaling.

---

Typing out the above got me thinking, we're modelling an abstract API, yet we're telling customers to interact with the concrete implementation of auto scaling. Is it me or does that seem odd? Perhaps we shouldn't be overwriting the annotations like we are? Has that been explored?

Reply:
The goal of the API design is to fit into the current product decision: we want to support both upstream autoscaler and autoscaler as an eks-a curated package. Just like any other package (load balancer, ingress controller, harbor), user can choose between installing it from upstream, or through eks-a gated, scanned curated package. To support upstream, we will need to expose the underlying CAPI `machinedeployment` min/max annotation to user so that the eks-a cluster controller reconciles and always applies it.

Now to expose the min/max annotation, there are two ways:
1. include it in cluster spec, as part of the worker node group config (chosen in this doc)
2. define an eks-a autoscaling CRD, map min/max to the worker node group. e.g.

```go
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: Autoscaler
metadata:
  name: cluster-1-autoscaler
spec:
  minCount: 1
  maxCount: 3
  workerNodeGroupRef:
    kind: VSphereMachineConfig
    name: md-0
```

The one benefit of this approach is to isolate cluster autoscaler config from the cluster spec. But it's more complicated and less straightforward API wise. One autoscaler CR maps to only one worker node group. Way more complicated implementation for both CLI and controller. And this still does not solve the confusion mentioned above. In the future even if we integrate with Karpenter, this CRD will not be used at all since Karpenter does not scale through node group and it has its own CRDs user can use.

So in general, the design API is the most straightforward, easy to implement solution. As for confusion, EKS follows similar pattern where user needs to create a cluster with autoscaling capability first:`eksctl create cluster --asg-access`, which will set up the EC2 ASG group, and the IAM permissions. Then manually install cluster autoscaler themselves from upstream to enable autoscaling function.

Reply:
If we use a CRD, we can reuse the CRD for different worker node groups right?